opentelemetry/src/context.rs

```diff
@@ -897,4 +979,171 @@ mod tests {
+
+        // Enter a suppressed scope
+        {
+            let _guard = Context::enter_telemetry_suppressed_scope();
+            assert!(Context::is_current_telemetry_suppressed());
+
+            // Test nested context - should not be suppressed
+            {
+                let _inner_guard = Context::new().attach();
+                assert!(!Context::is_current_telemetry_suppressed());
+            }
+
+            // Back to suppressed after inner guard is dropped
+            assert!(Context::is_current_telemetry_suppressed());
+        }
+
+        // Back to unsuppressed after outer guard is dropped
+        assert!(!Context::is_current_telemetry_suppressed());
+    }
+
+    #[test]
+    fn test_nested_suppression_scopes() {
+        // Ensure we start with a clean context
+        let _reset_guard = Context::new().attach();
+
+        // Default context should not be suppressed
+        assert!(!Context::is_current_telemetry_suppressed());
+
+        // First level suppression
+        {
+            let _outer = Context::enter_telemetry_suppressed_scope();
+            assert!(Context::is_current_telemetry_suppressed());
+
+            // Second level. This component is unaware of Suppression,
+            // and just attaches a new context. Since it is from current,
+            // it'll already have suppression enabled.
+            {
+                let _inner = Context::current().with_value(ValueA(1)).attach();
+                assert!(Context::is_current_telemetry_suppressed());
+            }
+
+            // Another scenario. This component is unaware of Suppression,
+            // and just attaches a new context, not from Current. Since it is
+            // not from current it will not have suppression enabled.
+            {
+                let _inner = Context::new().with_value(ValueA(1)).attach();
+                assert!(!Context::is_current_telemetry_suppressed());
+            }
+
+            // Still suppressed after inner scope
+            assert!(Context::is_current_telemetry_suppressed());
+        }
+
+        // Back to unsuppressed
+        assert!(!Context::is_current_telemetry_suppressed());
+    }
+
+    #[tokio::test]
+    async fn test_async_suppression() {
+        async fn nested_operation() {
+            assert!(Context::is_current_telemetry_suppressed());
+
+            let cx_with_additional_value = Context::current().with_value(ValueB(24));
+
+            async {
+                assert_eq!(
+                    Context::current().get::<ValueB>(),
+                    Some(&ValueB(24)),
+                    "Parent value should still be available after adding new value"
+                );
+                assert!(Context::is_current_telemetry_suppressed());
+
+                // Do some async work to simulate real-world scenario
+                sleep(Duration::from_millis(10)).await;
+
+                // Values should still be available after async work
```

Comment:
`#[tokio::test]` would be using the current-threaded runtime. Could we test this for the multi-threaded runtime as well? I suspect we would have to update the sleep call with the context for that.
```rust
sleep(Duration::from_millis(10)).with_context(Context::current()).await;
```

Reply:
since we are not doing anything *inside* sleep() call, what is the need to invoking it with_context? If it was any other async method, then we should invoke with with_context(), so that within the body of that method, we have correct context. I am not following why we need to sleep with context....

Reply:
My bad. I was concerned about how context gets propagated across the sleep await point. I forgot about the `Future` implementation for `WithContext` which ensures that things work correctly https://github.com/open-telemetry/opentelemetry-rust/blob/f3e93a09ea9059487c4a8a7e402fce00cc56aa34/opentelemetry/src/context/future_ext.rs#L13-L18